### DISCUSSION TITLE
How to keep the connection alive to MongoDB?

### DISCUSSION THREAD

**Discussion by kaykhan** [prisma/prisma#15284] [Q&A] [answered]
I have a cron job which has 2 stages. 1st stage it fetches data, it takes around 3 or 4 minutes to fetch the data and then the 2nd stage it makes a single bulk write. 

however i get the following error:

> RawDatabaseError { code: \"unknown\", message: \"Broken pipe (os error 32)\" }

In this case it seems like prisma had established a connection with mongodb server at some point, and then that connection has been broken so when the logic goes to insert, it throws the above error:

Is there some way to keep the connection to mongodb alive?

Im using the following mongodb connection url options

```
&retryWrites=true&w=majority&connectTimeoutMS=300000&serverSelectionTimeoutMS=300000&socketTimeoutMS=300000&minPoolSize=10
```

https://stackoverflow.com/questions/62718415/what-is-the-use-of-nodejs-mongodb-driver-keepalive-option

**Comment by janpio** [comment]
I don't think this should throw such an error. If you can reproduce this, feel free to open a bug issue.

**Reply by kaykhan** [reply]
Why do you say that, is there something underlining in prisma that holds onto (or retries) the mongodb connection that you can reference me to? 

becuase my theory is If you look at the referenced link, i believe this is what is happening - `Preventing disconnection due to network inactivity`. I belive during the first phase of my cron -fetching data from elsewhere and doing nothing with mongodb- the connection is being closed to mongodb. 

if i increase the idle timeout of the ELB (60seconds to 300 seconds) which sits infront of my mongodb, it works ( no broken pipe error). (increasing the idle timeout of the elb is not the appropriate solution here, the mongodb should be kept alive or retried)

Which tells me in the first scenario the connection is disconnected, and prisma never reconnects. (in nodejs i init prisma as a signleton)

**Reply by janpio** [reply]
I say that because Prisma should not throw an error, but should do the right thing (keep the connection alive if possible, reopene the connection if not) so this never happens.

**Reply by kaykhan** [reply]
> I say that because Prisma should not throw an error, but should do the right thing (keep the connection alive if possible, reopene the connection if not) so this never happens.

Yes i would like for it to keepAlive or retry on its own, but unfortunately its not. I also can't find any documentation which allows me to configure this.

Its also a networking problem which is hard for me to give you a reproduction of. But i can explain the architecture.

1. Sestup MongoDB instance exposed via an AWS ELB
2. Setup a node.js application using prisma. Initialize prisma as a singleton which connect to the ELB via TCP.
3. Create a main function.
   - create a setTimeout function of (2 mins, greater than the default elb idle timeout) (this recreates doing nothing)
   - try to make a query to mongodb after the period of doing nothing. 
   
   
( you can see why its difficult for me to make a reproduction because in docker we will not have this idle timeout as we have in load balnacer).

**Reply by kaykhan** [reply]
If there is someway via prisma that allowed me to set this boolean (keepAlive) for the mongodb driver https://mongodb.github.io/node-mongodb-native/4.9/interfaces/ConnectionOptions.html#keepAlive i believe it would help me. https://www.mongodb.com/docs/drivers/node/current/faq/#what-does-the--keepalive--setting-do-

**Reply by kaykhan** [reply]
@janpio if i use the prisma middleware to add my own retry mechanism, i can get it to work. Really confused why this happens

```
PrismaService.$use(async (params, next) => {
    // Manipulate params here
    console.log(params);
    let c = 0;
    let response;
    while (c < 5) {
        console.log("count", c);
        try {
            const result = await next(params);
            response = result;
            break;
        } catch (err) {
            c++;
            console.log(err);
            response = err;
        }
    }
    return response;
});
```

**stdout**
```
params {
  args: {
    command: {
      insert: 'salesByContractAddress',
      documents: [Array],
      ordered: false
    }
  },
  dataPath: [],
  runInTransaction: false,
  action: 'runCommandRaw',
  model: undefined
}
PrismaClientUnknownRequestError:
Invalid `prisma.$runCommandRaw()` invocation:


Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: RawDatabaseError { code: "unknown", message: "Broken pipe (os error 32)" } })
    at RequestHandler.handleRequestError (/home/kay/checkpoint/app/core/node_modules/@prisma/client/runtime/index.js:29913:13)
    at RequestHandler.request (/home/kay/checkpoint/app/core/node_modules/@prisma/client/runtime/index.js:29892:12)
    at async /home/kay/checkpoint/appcore/src/services/mongodb.service.ts:17:28
    at async Proxy._request (/home/kay/checkpoint/app/core/node_modules/@prisma/client/runtime/index.js:30864:16)
    at async InsertSalesByContractAddress (/home/kay/checkpoint/app/core/src/models/sales-by-contract-address/sales-by-contract-address.repo.ts:24:35)
    at async LoadData (/home/kay/checkpoint/napp/cron/src/daily.ts:97:39)
    at async Object.Process [as fn] (/home/kay/checkpoint/app/cron/src/daily.ts:80:35)
    at async Pipeline.promise (/home/kay/checkpoint/app/cron/src/modules/pipeline.module.ts:65:26) {
  clientVersion: '4.3.1'
}
params {
  args: {
    command: {
      insert: 'salesByContractAddress',
      documents: [Array],
      ordered: false
    }
  },
  dataPath: [],
  runInTransaction: false,
  action: 'runCommandRaw',
  model: undefined
}
params {
  args: {
    command: {
      insert: 'salesByContractAddress',
      documents: [Array],
      ordered: false
    }
  },
  dataPath: [],
  runInTransaction: false,
  action: 'runCommandRaw',
  model: undefined
}
params {
  args: {
    command: {
      insert: 'salesByContractAddress',
      documents: [Array],
      ordered: false
    }
  },
  dataPath: [],
  runInTransaction: false,
  action: 'runCommandRaw',
  model: undefined
}
params {
  args: {
    command: {
      insert: 'salesByContractAddress',
      documents: [Array],
      ordered: false
    }
  },
  dataPath: [],
  runInTransaction: false,
  action: 'runCommandRaw',
  model: undefined
}
{"level":"info","message":{"nft-transpose":{"kind":"LoadData","response":{"n":6,"ok":1}}},"timestamp":"2022-09-13T12:42:31.947Z"}
params {
  args: {
    where: { address: '0x28472a58a490c5e09a238847f66a68a47cc76f0f' },
    data: {
      address: '0x28472a58a490c5e09a238847f66a68a47cc76f0f',
      last_execution: '2022-09-13T00:00:00.000Z',
      last_update: '2022-09-13T00:00:00.000Z'
    }
  },
  dataPath: [],
  runInTransaction: false,
  action: 'update',
  model: 'Contract'
}
prisma:query db.contracts.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $eq: [ "$address", "0x28472a58a490c5e09a238847f66a68a47cc76f0f", ], }, { $ne: [ "$address", "$$REMOVE", ], }, ], }, ], }, }, }, { $project: { _id: 1, }, }, ])
prisma:query db.contracts.updateMany({ _id: { $in: [ ObjectId("631f373d96ffad66a9a37005"), ], }, }, [{ $set: { address: { $literal: "0x28472a58a490c5e09a238847f66a68a47cc76f0f", }, }, }{ $set: { last_update: { $literal: DateTime("2022-09-13 0:00:00.0 +00:00:00"), }, }, }{ $set: { last_execution: { $literal: DateTime("2022-09-13 0:00:00.0 +00:00:00"), }, }, }])
prisma:query db.contracts.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $eq: [ "$_id", ObjectId("631f373d96ffad66a9a37005"), ], }, { $ne: [ "$_id", "$$REMOVE", ], }, ], }, ], }, }, }, { $project: { _id: 1, network: 1, address: 1, active: 1, last_update: 1, last_execution: 1, }, }, ])
```

**Reply by janpio** [reply]
As I wrote before: If you can reproduce this, feel free to open a bug issue.

**Comment by kaykhan** [comment, marked as answer]
I managed to found a suitable workaround until this can be addressed directly in prisma.

The workaround is to retry requests using prisma middleware. I specifically look for the error `PrismaClientUnknownRequestError` to retry those.

```
PrismaClient.$use(async (params, next) => {
    let retries = 1;
    const maxRetries = 5;

    do {
        try {
            const response = await next(params);
            return response;
        } catch (err) {
            Logger.warn({
                kind: "prisma",
                retry: retries,
                property: err,
                stack_trace: err.stack,
            });
            if (err instanceof Prisma.PrismaClientUnknownRequestError) {
                retries++;
                continue;
            }
            throw err;
        }
    } while (retries < maxRetries);
});
```
